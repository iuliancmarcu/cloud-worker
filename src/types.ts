export type TaskType = 'FIBO' | 'FACT';

export type Task = {
  id: string;
  type: TaskType;
  payload: object;
};

export type CompletedTask = Task & {
  result: object;
};

export type FiboTask = Task & {
  type: 'FIBO';
  payload: {
    n: number;
  };
};

export type FactTask = Task & {
  type: 'FACT';
  payload: {
    n: number;
  };
};
