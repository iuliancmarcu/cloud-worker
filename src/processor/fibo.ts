import { FiboTask, CompletedTask } from '../types';

export async function process(task: FiboTask): Promise<CompletedTask> {
  const fibo = fibonacci(task.payload.n);

  return {
    ...task,
    result: {
      value: fibo,
    },
  };
}

function fibonacci(n: number): number {
  if (n <= 2) {
    return 1;
  }

  return fibonacci(n - 2) + fibonacci(n - 1);
}
