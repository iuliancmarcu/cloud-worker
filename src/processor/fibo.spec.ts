import * as fibo from './fibo';
import { FiboTask } from '../types';

describe('Processor: fibo', () => {
  it('processes a FIBO task and resolves with a CompletedTask', () => {
    const task: FiboTask = {
      id: 'random-id',
      type: 'FIBO',
      payload: {
        n: 10,
      },
    };

    expect(fibo.process(task)).resolves.toMatchSnapshot();
  });
});
