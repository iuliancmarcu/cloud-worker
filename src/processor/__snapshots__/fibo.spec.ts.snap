// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Processor: fibo processes a FIBO task and resolves with a CompletedTask 1`] = `
Object {
  "id": "random-id",
  "payload": Object {
    "n": 10,
  },
  "result": Object {
    "value": 55,
  },
  "type": "FIBO",
}
`;
